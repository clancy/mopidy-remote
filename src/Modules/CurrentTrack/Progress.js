import React, { Component, PropTypes } from 'react';
import {
  View,
  StyleSheet,
  Text,
  ProgressViewIOS
} from 'react-native';

const padZero = (num, size) => {
  var s = num+"";
  while (s.length < size) s = "0" + s;
  return s;
};

export default Progress = React.createClass({
  propTypes: {
    length: PropTypes.number,
    position: PropTypes.number,
    playing: PropTypes.bool.isRequired
  },

  getTrackProgress: (length, position) => {
    return position / length;
  },

  msToMinAndSec: (ms) => {
    if(ms === undefined){
      return '0:00'
    }
    var x = Math.trunc(ms / 1000)
    var seconds = x % 60
    x = Math.trunc(x / 60)
    var minutes = x % 60
    return `${minutes}:${padZero(seconds,2)}`
  },
  getInitialState: function() {
    return {
      position: 0
    };
  },
  tick: function() {
    this.setState({position: this.state.position + 1000});
    if (this.state.position >= this.props.length) {
      clearInterval(this.interval);
    }
  },
  componentDidMount: function() {
    this.setState({ position: this.props.position });

    clearInterval(this.interval);
    if(this.props.playing){
      this.interval = setInterval(this.tick, 1000);
    }
  },
  componentWillReceiveProps(nextProps) {
    if(this.props == nextProps){
      return;
    }
    this.setState({ position: nextProps.position });

    clearInterval(this.interval);
    if(nextProps.playing){
      this.interval = setInterval(this.tick, 1000);
    }
  },
  componentWillUnmount: function() {
    clearInterval(this.interval);
  },
  render() {
    return (
      <View style={styles.container} >
        <View style={styles.timeContainer}>
          <Text style={styles.timeText}>{this.msToMinAndSec(this.state.position)}</Text>
        </View>
        <ProgressViewIOS
          style={styles.progressView}
          progress={this.getTrackProgress(this.props.length, this.state.position)}
          progressTintColor='green'/>
        <View style={styles.timeContainer}>
          <Text style={styles.timeText}>{this.msToMinAndSec(this.props.length)}</Text>
        </View>
      </View>
    );
  }
});

var styles = StyleSheet.create({
  container: {
    flex: .5,
    backgroundColor: 'black',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  timeContainer: {
    flex: 15,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  timeText: {
    color: 'white'
  },
  progressView: {
    flex: 70
  }
});
