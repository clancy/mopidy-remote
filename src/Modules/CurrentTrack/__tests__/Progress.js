import 'react-native';
import React from 'react';
import Progress from '../Progress';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
describe('Progress tests', () => {
  test('renders Progress correctly with 0 lenght', () => {
    const tree = renderer.create(
      <Progress
        length={0}
        position={0}
        playing={true}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders Progress correctly with length and position', () => {
    const tree = renderer.create(
      <Progress
        length={4000}
        position={60000}
        playing={false}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
