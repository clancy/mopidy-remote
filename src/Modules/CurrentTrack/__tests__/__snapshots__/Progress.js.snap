// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress tests renders Progress correctly with 0 lenght 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 0.5,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 15,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
        }
      }
    >
      0:00
    </Text>
  </View>
  <RCTProgressView
    progress={0}
    progressTintColor="green"
    style={
      Array [
        Object {
          "height": 2,
        },
        Object {
          "flex": 70,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 15,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
        }
      }
    >
      0:00
    </Text>
  </View>
</View>
`;

exports[`Progress tests renders Progress correctly with length and position 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "flex": 0.5,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 15,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
        }
      }
    >
      1:00
    </Text>
  </View>
  <RCTProgressView
    progress={15}
    progressTintColor="green"
    style={
      Array [
        Object {
          "height": 2,
        },
        Object {
          "flex": 70,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 15,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
        }
      }
    >
      0:04
    </Text>
  </View>
</View>
`;
